packages/annotations/src/format/annotation.js

```diff
@@ -52,31 +57,104 @@ export function removeAnnotations( record ) {
 	return removeFormat( record, 'core/annotation', 0, record.text.length );
 }
 
+/**
+ * Retrieves the positions of annotations inside an array of formats.
+ *
+ * @param {Array} formats Formats with annotations in there.
+ * @return {Object} ID keyed positions of annotations.
+ */
+function retrieveAnnotationPositions( formats ) {
+	const positions = {};
+
+	formats.forEach( ( characterFormats, i ) => {
+		characterFormats = characterFormats || [];
```

Comment:
This should always be an array, so the casting seems unnecessary?

Reply:
It can be `undefined` right, for when no formatting is present on that character?

Reply:
It's a sparse array, there's no indices set as undefined, they're just not there. `forEach` will only loop over any indices that are set, which is only arrays.

Reply:
Try:

```js
[ , [] ].forEach( ( { forEach } ) => {} );
[ undefined, [] ].forEach( ( { forEach } ) => {} );
```